In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm #statistic library
import seaborn
seaborn.set()

In [2]:
#handle error because of numpy
import warnings
warnings.filterwarnings("ignore", message="numpy.dtype size changed")
warnings.filterwarnings("ignore", message="numpy.ufunc size changed")

In [3]:
df = pd.read_csv('udemy_data/1.02. Multiple linear regression.csv')
df.head()

SAT   GPA  Rand 1,2,3
0  1714  2.40           1
1  1664  2.52           3
2  1760  2.54           3
3  1685  2.74           3
4  1693  2.83           2

<h3>Create Multiple Regression</h3>
<p>The different between simple regression and multiple regression is the number of feature used as independent variable</p>
<p>GPA = b0 + b1.SAT + b2.Rand + epsilon</p>

In [4]:
y = df['GPA'] #dependent variable 
x1 = df[['SAT','Rand 1,2,3']] #independent variable

In [5]:
x = sm.add_constant(x1) #we always need state model.add constant to define which the feature use for training
results = sm.OLS(y,x).fit() #training process
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    GPA   R-squared:                       0.407
Model:                            OLS   Adj. R-squared:                  0.392
Method:                 Least Squares   F-statistic:                     27.76
Date:                Fri, 18 Sep 2020   Prob (F-statistic):           6.58e-10
Time:                        23:06:35   Log-Likelihood:                 12.720
No. Observations:                  84   AIC:                            -19.44
Df Residuals:                      81   BIC:                            -12.15
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.2960      0.417      0.710      0.480      -0.533       1.125
SAT            0.0017      0.000      7.432      0.000       0.001       0.002
Rand 1,2,3    -0.0083      0.027     -0.304      0.762      -0.062       0.046
==============================================================================
Omnibus:                       12.992   Durbin-Watson:                   0.948
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               16.364
Skew:                          -0.731   Prob(JB):                     0.000280
Kurtosis:                       4.594   Cond. No.                     3.33e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.33e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- If we compare with the first model (1. Simple Linear Regression),R-square is bigger (from 0.406 to 0.407), but decrease Adj.R square(from 0.399 to 0.392). We have add information (r square) but lose value (adj r)
- Look at the Rand 123 feature, The value of coef is minus (-), so it will change the formula to b0+b1x1+(-b2)x2. 
- Look at the coef table, P value 0.762 of Rand feature is insignificant because we want p-val < 0.05, so we have drop the feature. Dropping useless variable is important
- Adjusted R-square is the basic for comparing regression model but that isn't quite true.
- Based on Prob(F statistic), the lower of f-statistic the closer to non-significant model | before 7.20e-11 increase 6.58e-10.